pkg/storer/compact.go

```diff
@@ -0,0 +1,192 @@
+// Copyright 2023 The Swarm Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package storer
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"path"
+	"sort"
+	"time"
+
+	"github.com/ethersphere/bee/pkg/cac"
+	"github.com/ethersphere/bee/pkg/log"
+	"github.com/ethersphere/bee/pkg/sharky"
+	"github.com/ethersphere/bee/pkg/soc"
+	"github.com/ethersphere/bee/pkg/storage"
+	"github.com/ethersphere/bee/pkg/storer/internal/chunkstore"
+	"github.com/ethersphere/bee/pkg/swarm"
+	"golang.org/x/sync/errgroup"
+)
+
+// Compact minimizes sharky disk usage by, using the current sharky locations from the storer,
+// relocating chunks starting from the end of the used slots to the first available slots.
+func Compact(ctx context.Context, basePath string, opts *Options, validate bool) error {
+
+	logger := opts.Logger
+
+	store, err := initStore(basePath, opts)
+	if err != nil {
+		return fmt.Errorf("failed creating levelDB index store: %w", err)
+	}
+	defer func() {
+		if err := store.Close(); err != nil {
+			logger.Error(err, "failed closing store")
+		}
+	}()
+
+	sharkyRecover, err := sharky.NewRecovery(path.Join(basePath, sharkyPath), sharkyNoOfShards, swarm.SocMaxChunkSize)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if err := sharkyRecover.Close(); err != nil {
+			logger.Error(err, "failed closing sharky recovery")
+		}
+	}()
+
+	if validate {
+		logger.Info("performing chunk validation before compaction")
+		validationWork(logger, store, sharkyRecover)
+	}
+
+	logger.Info("starting compaction")
+
+	n := time.Now()
+
+	for shard := 0; shard < sharkyNoOfShards; shard++ {
+
+		select {
+		case <-ctx.Done():
+			return errors.Join(ctx.Err(), sharkyRecover.Save())
+		default:
+		}
+
+		items := make([]*chunkstore.RetrievalIndexItem, 0, 1_000_000)
+		// we deliberately choose to iterate the whole store again for each shard
+		// so that we do not store all the items in memory (for operators with huge localstores)
+		_ = chunkstore.Iterate(store, func(item *chunkstore.RetrievalIndexItem) error {
+			if item.Location.Shard == uint8(shard) {
+				items = append(items, item)
+			}
+			return nil
+		})
+
+		sort.Slice(items, func(i, j int) bool {
+			return items[i].Location.Slot < items[j].Location.Slot
+		})
+
+		lastUsedSlot := items[len(items)-1].Location.Slot
+		slots := make([]*chunkstore.RetrievalIndexItem, lastUsedSlot+1) // marks free and used slots
+		for _, l := range items {
+			slots[l.Location.Slot] = l
+		}
+
+		// start begins at the zero slot. The loop below will increment the position of start until a free slot is found.
+		// end points to the last slot, and the loop will decrement the position of end until a used slot is found.
+		// Once start and end point to free and used slots, respectively, the swap of the chunk location will occur.
+		start := uint32(0)
+		end := lastUsedSlot
+
+		for start < end {
+			if slots[start] == nil { // free
```

Comment:
you need to make sure `end` is always set to a used slot otherwise you end up with suboptimal truncation: 

imagine
```
0 1 2 3 4 5 6 7 8 9 ........
+ + + + + + + - - + - - - - - 
````

after the first move at `start = 7; end = 9`, the loop terminates at `start = 8; end = 8` instead of `end = 7` 

i suggest:
```
end := lastUsedSlot
for start  := uint(32); start < end;  {
      if slots[end] == nil {
          end--
          continue
      }
      if slots[start] != nil {
         start++
         continue
     }
     MOVE
     start++
     end--
}
```